sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/compiler/QueryGenerator.kt

```diff
@@ -34,59 +34,65 @@ abstract class QueryGenerator(private val query: BindableQuery) {
   protected fun executeBlock(): CodeBlock {
     val result = CodeBlock.builder()
 
-    val maxIndex = query.arguments.filter { it.type.bindArg?.isArrayParameter() != true }.map { it.index }.max() ?: 0
     val precedingArrays = mutableListOf<String>()
     val bindStatements = CodeBlock.builder()
     val replacements = mutableListOf<Pair<IntRange, String>>()
     val argumentCounts = mutableListOf<String>()
 
     var needsFreshStatement = false
 
-    query.arguments.filterNot { it.type.bindArg?.isArrayParameter() == true }.size.let {
-      if (it != 0) {
-        argumentCounts.add(it.toString())
+    val positionToArgument = mutableListOf<Triple<Int, BindableQuery.Argument, SqlBindExpr?>>()
+    query.arguments.forEach { argument ->
+      if (argument.bindArgs.isNotEmpty()) {
+        argument.bindArgs.forEach { bindArg ->
+          positionToArgument.add(Triple(bindArg.node.textRange.startOffset, argument, bindArg))
```

Comment:
while splitting out this method should probably have a more semantic data class for these

Reply:
Yep, agree. `Triple` is gross, this wants a data class